src/query_plan/operation.rs

```diff
@@ -0,0 +1,435 @@
+    }
+}
+
+impl From<NormalizedSelectionSet> for SelectionSet {
+    fn from(val: NormalizedSelectionSet) -> Self {
+        SelectionSet {
+            ty: val.ty.clone(),
+            selections: flatten_selections(&val.selections),
+        }
+    }
+}
+
+/// Converts vec of Selections to a map of NormalizedSelections
+///
+/// Expands all named fragments and merge duplicate selections.
+fn normalize_selections(
+    selections: &Vec<Selection>,
+    fragments: &IndexMap<Name, Node<Fragment>>,
+) -> NormalizedSelectionMap {
+    let mut normalized = NormalizedSelectionMap::new();
+    for selection in selections {
+        match selection {
+            Selection::Field(field) => {
+                let expanded_selection_set =
+                    normalize_selections(&field.selection_set.selections.to_owned(), fragments);
+
+                if let NormalizedSelection::NormalizedField(field_entry) = normalized
+                    .entry(field.into())
+                    .or_insert(NormalizedSelection::NormalizedField(Node::new(
+                        NormalizedField {
+                            definition: field.definition.clone(),
+                            alias: field.alias.clone(),
+                            name: field.name.clone(),
+                            arguments: field.arguments.clone(),
+                            directives: field.directives.clone(),
+                            selection_set: NormalizedSelectionSet {
+                                ty: field.selection_set.ty.clone(),
+                                selections: IndexMap::new(),
+                            },
+                        },
+                    )))
+                {
+                    let merged_selections = merge_selections(
+                        &field_entry.selection_set.selections,
+                        &expanded_selection_set,
+                    );
+                    let mut selection_set = field_entry.selection_set.clone();
+                    selection_set.selections = merged_selections;
+                    field_entry.make_mut().selection_set = selection_set;
+                    // field_entry.selection_set.selections = merged_selections;
+                }
+            }
+            Selection::FragmentSpread(named_fragment) => {
+                if let Some(fragment) = fragments.get(&named_fragment.fragment_name) {
+                    let expanded_selection_set = normalize_selections(
+                        &fragment.selection_set.selections.to_owned(),
+                        fragments,
+                    );
+                    normalized = merge_selections(&normalized, &expanded_selection_set);
+                } else {
+                    // no fragment found
+                }
+            }
+            Selection::InlineFragment(inline_fragment) => {
+                let expanded_selection_set = normalize_selections(
+                    &inline_fragment.selection_set.selections.to_owned(),
```

Comment:
Similarly here, you won't need the `to_owned()`.
```suggestion
                    &inline_fragment.selection_set.selections,
```